### DISCUSSION TITLE
Instantiation of patterns: generation from a pattern file, or pre-written template files?

### DISCUSSION THREAD

**Discussion by willosborne** [finos/architecture-as-code#71] [Ideas]
As we think about instantiating patterns for the CALM CLI proposed in Issue #53, we have a couple of options. Wanted to get people's opinions on how to approach this.

What do people think?

## Option 1: Generate an instantiation by inspecting the pattern file
This is a flexible approach that takes a pattern, traverses the spec and generates a JSON file as output.
For example, if it sees a property with 
```
"const": "Value here"
```
... it would simply insert this into the generated text.
If it saw a type with no value specified a placeholder such as `{{ PROPERTY_NAME }}` would be inserted so we can easily template later.

One library that looks promising here is the [`json-schema-faker` library here](https://github.com/json-schema-faker/json-schema-faker?tab=readme-ov-file) that takes a schema and generates a JSON document. Our use case isn't exactly the same from a quick look at the docs but this is worth investigating.

## Option 2: Pattern files come with a template for instantiating it
In this model a pattern is two files: the pattern itself as a JSON schema, and an instantiation of the pattern with {{ PLACEHOLDER }} values in all the slots that needs to be filled.

This is a lot simpler, but requires two files. This would make fetching the pattern from some kind of schema repository more complex. Pattern files could also potentially reference their template.

It also requires maintenance to ensure there is no drift between the instantiation and the template.

**Comment by willosborne** [comment]
@rocketstack-matt @jpgough-ms Interested in your thoughts here

**Reply by rocketstack-matt** [reply]
I would lean towards option 1, I think option 2 will result in us effectively needing to do option 1 purely as a validator to ensure the template is correct, without that we would very likely have templates that didn't match or drifted away from the pattern.
